File: frontend/src/pages/sanvaad/varta/waWindow.js
/**
 * waWindow.js — Meta's 24-hour customer service window.
 *
 * `06-sanvaad-varta.md` §9: "Outside a 24-hour window from the customer's last
 * message, Meta rejects free-form text — only approved templates go through.
 * WAChat offers a plain text field regardless, so the send fails and surfaces as
 * a generic error toast. The composer must know the window state: countdown
 * while open, template picker when closed."
 *
 * `06` §4 wants a new `GET /whatsapp/conversations/:id/window` returning
 * `{open, expires_at}`. The backend is not this module's to edit — and it does
 * not need to be for correctness here: `varta_messages` carries `direction` and
 * `created_at`, and the conversation's message page is already fetched, so the
 * window is the newest inbound message plus 24 hours. The endpoint is still
 * worth adding, because this derivation only sees the newest page (50 messages);
 * a conversation with more than 50 outbound messages since the last inbound one
 * reads as "never opened", which is the safe direction to be wrong in — it
 * offers templates rather than a send that Meta will reject.
 *
 * ── The endpoint now exists ─────────────────────────────────────────────────
 *
 * `GET /api/v1/whatsapp/conversations/:id/window` was added, and so was the
 * thing that actually matters: `POST .../messages` REFUSES a free-form send
 * outside the window rather than storing it and reporting 201. A rule enforced
 * only by the control that offers it is not enforced — the composer hid the
 * text field, and a stale tab, a retry or curl sent through it anyway, at which
 * point Meta rejected the message at its edge and our record of the
 * conversation stopped matching the customer's.
 *
 * Both derivations are kept and they do different jobs. `fromServer` is the
 * authority — it reads MAX(created_at) over every inbound row, not a page of
 * fifty. `windowState` is the fallback while that request is in flight or after
 * it fails, so the composer never renders in an unknown state, and it is the
 * one the tests exercise for the arithmetic.
 */

export const WINDOW_MS = 24 * 60 * 60 * 1000;

/** `{open, expiresAt, remainingMs, everInbound}` from a message list. */
export function windowState(messages, now = Date.now()) {
  let lastInbound = 0;
  for (const m of messages) {
    if (m?.direction !== 'inbound') continue;
    const t = new Date(m.created_at || 0).getTime();
    if (t > lastInbound) lastInbound = t;
  }
  if (!lastInbound) return { open: false, expiresAt: null, remainingMs: 0, everInbound: false };
  const expiresAt = lastInbound + WINDOW_MS;
  const remainingMs = expiresAt - now;
  return { open: remainingMs > 0, expiresAt, remainingMs, everInbound: true };
}

/**
 * The same `{open, expiresAt, remainingMs, everInbound}` shape, from the
 * server's `{open, expires_at, remaining_seconds, ever_inbound}`.
 *
 * `open` is recomputed from `expires_at` against the caller's clock rather than
 * trusted from the payload. The payload's own `open` was true at the instant
 * the request was served; between a five-second poll and a user reading the
 * screen, a window with forty seconds left expires while the banner still says
 * it is open. The timestamp is the fact; the boolean is that fact at a moment
 * that has passed.
 *
 * Returns `null` — not a closed window — when the payload cannot be read.
 * A null tells the caller to fall back to `windowState`; a closed window would
 * silently replace the composer for a conversation that is fine.
 */
export function fromServer(payload, now = Date.now()) {
  if (!payload || typeof payload !== 'object') return null;
  if (!payload.ever_inbound) {
    return { open: false, expiresAt: null, remainingMs: 0, everInbound: false };
  }
  const expiresAt = new Date(payload.expires_at || 0).getTime();
  if (!Number.isFinite(expiresAt) || expiresAt === 0) return null;
  const remainingMs = expiresAt - now;
  return { open: remainingMs > 0, expiresAt, remainingMs, everInbound: true };
}

/** "7h 20m left" — coarse on purpose; a ticking second hand on a 24-hour
 *  countdown is motion with no information in it. */
export function formatRemaining(ms) {
  if (ms <= 0) return 'closed';
  const mins = Math.floor(ms / 60000);
  const h = Math.floor(mins / 60);
  const m = mins % 60;
  if (h >= 1) return `${h}h ${m}m left`;
  return `${m}m left`;
}

File: backend/services/wa_window.py
"""
wa_window.py — Meta's 24-hour customer service window, decided on the SERVER.

`frontend/src/pages/sanvaad/varta/waWindow.js` already models this window, and
its own header says why that is not enough:

    "The endpoint is still worth adding, because this derivation only sees the
     newest page (50 messages); a conversation with more than 50 outbound
     messages since the last inbound one reads as 'never opened'."

That is the benign half of the problem. The other half is that a UI-only rule is
not a rule at all. `POST /conversations/{id}/messages` took `{content, type}`
from anyone who could reach the route and wrote it straight into
`varta_messages` with no reference to the window — so a closed conversation
could be sent free-form text by curl, by a stale tab whose `windowState` was
computed an hour ago, or by any client that simply did not run our JavaScript.
Meta rejects those at the edge, which means the product's own record of what it
sent diverges from what the customer received, and a Business Account that keeps
attempting them gets rate-limited and eventually flagged.

So the window is derived here, from the whole conversation rather than a page of
it, and the send route refuses on it. The frontend keeps its copy: it is what
turns the composer into a template picker BEFORE the user types, which is a
different job from refusing, and both are wanted.

WHAT THE WINDOW IS
------------------
Meta opens a 24-hour "customer service window" at each INBOUND message. Inside
it, free-form messages of any type are allowed. Outside it — including a
conversation where the customer has never written first — only a template Meta
has APPROVED may be sent.

The clock is the newest inbound message, not the newest message: an outbound
reply does not extend the window, which is the single most common
misunderstanding of this rule and the reason `direction='inbound'` is in the
WHERE rather than an ORDER BY over everything.
"""
from datetime import datetime, timedelta, timezone
from typing import Optional

WINDOW_SECONDS = 24 * 60 * 60


def state_from(last_inbound: Optional[datetime], now: Optional[datetime] = None) -> dict:
    """`{open, expires_at, remaining_seconds, ever_inbound}` from one timestamp.

    Pure, so the arithmetic can be tested without a database. `now` is injected
    for the same reason.

    A naive timestamp is read as UTC rather than rejected: asyncpg returns
    TIMESTAMPTZ as aware, but a seed, a fixture or a driver change that hands
    over a naive value must not make this raise inside a send.
    """
    now = now or datetime.now(timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)

    if not last_inbound:
        # Never written to us. Meta treats this as closed, not as open-by-default
        # — a business cannot open a conversation with free-form text.
        return {
            "open": False,
            "expires_at": None,
            "remaining_seconds": 0,
            "ever_inbound": False,
        }

    if last_inbound.tzinfo is None:
        last_inbound = last_inbound.replace(tzinfo=timezone.utc)

    expires_at = last_inbound + timedelta(seconds=WINDOW_SECONDS)
    remaining = int((expires_at - now).total_seconds())
    return {
        "open": remaining > 0,
        "expires_at": expires_at.isoformat(),
        "remaining_seconds": max(0, remaining),
        "ever_inbound": True,
    }


async def window_state(pool, conv_id: str, org_id: str) -> dict:
    """The window for one conversation, over ALL of its inbound messages.

    `org_id` is in the WHERE as well as `conversation_id`. The conversation has
    already been proved to belong to the org by the caller, so this is
    redundant — and it stays, because the day someone reuses this helper without
    that proof is the day the redundancy is the only thing scoping it.
    """
    row = await pool.fetchrow(
        """
        SELECT MAX(created_at) AS last_inbound
        FROM public.varta_messages
        WHERE conversation_id = $1::uuid
          AND org_id = $2::uuid
          AND direction = 'inbound'
        """,
        conv_id,
        org_id,
    )
    last_inbound = row["last_inbound"] if row else None
    return state_from(last_inbound)
